CSM Threats eBPF-free agent guide (#25105)
DataDog/documentation: content/en/security/cloud_security_management/guide/eBPF-free-agent.md

---
title: Threat Detection for Linux Without eBPF Support
disable_toc: false
---

This guide describes how to set up the CMS Threats eBPF-less solution for eBPF disabled environments such as AWS Fargate. The eBPF-less solution uses a ptrace-based Datadog agent.

This guide also describes some advantages of the ptrace solution.

<div class="alert alert-info">Threat Detection for Linux Without eBPF Support is in Preview. Reach out to your Datadog representative to sign up.</div>

## Summary of agent options

CMS Threats includes two agent options for threat detection and response:

- eBPF solution
- eBPF-less solution with ptrace: This version is only available where eBPF is not (Linux kernel versions 3.4 to 4.14).

{{% collapse-content title="eBPF solution" level="h4" %}}

Datadog has built all its security products around [eBPF (extended Berkeley Packet Filter)][1]. Some of the benfits of eBPF are:

> - eBPF improves safety by validating each program through the Linux kernel verifier. This ensures that a program can’t crash, fall into infinite loops, harm the system, etc.

Reviewer: ```suggestion
- eBPF improves safety by validating each program through the Linux kernel verifier. This ensures that a program can’t crash, fall into infinite loops, or harm the system.
```

- eBPF is JIT (Just In Time) compiled and the output bytecode is executed on a eBPF VM sandbox. This prevents any kernel crash and provides competitive performance.
- Easy to debug and maintain, can dynamically load programs, and has access to all information needed to trace the user space.

> The Datadog eBPF agent code is [fully open source][2].

Reviewer: ```suggestion
The Datadog eBPF Agent code is [fully open source][2].
```

{{% /collapse-content %}} 

{{% collapse-content title="eBPF-less solution with ptrace" level="h4" %}}
Some environments use instances with old kernels that do not have eBPF at all. The ptrace solution is provided for these environments.

> The following features are not available in the eBPF-less agent:

Reviewer: ```suggestion
The following features are not available in the eBPF-less Agent:
```

- Security profiles, providing:
  - Anomaly detection
  - Auto-suppression of normal behavior for signal triaging
  - Malware detection
- Network detections

> <div class="alert alert-info">The current implementation supports amd64 and arm64 architecture and ABIs but can be extended to 32-bit ABIs.</div>

Reviewer: ```suggestion
<div class="alert alert-info">The current implementation supports amd64 and arm64 architecture and ABIs, but can be extended to 32-bit ABIs.</div>
```

- Precise process control: ptrace provides detailed inspection of memory and registers, safeguarding critical application workloads. This granular visibility is essential for identifying sophisticated threats. The Datadog procfs (Process Filesystem) scanner monitors all system-wide executions, enabling the surgical termination of malicious processes. Together, these tools protect from malicious activity.
- Operational stability: Operating in user space, ptrace avoids the complexities and risks of kernel space, providing a safer and more manageable approach. In the event of a failure, a ptrace-based agent defaults to a fail-open state at the OS layer, keeping the system unaffected, even if the application hangs.
- Performance efficiency: Recent benchmarks conducted by Datadog's engineering team demonstrate that the Datadog ptrace-based implementation shows comparable performance to kernel-based solutions. Specifically, it introduces only a minimal overhead of around 3% for PostgreSQL workloads and negligible impacts for Redis operations, making it very efficient for most use cases.

> - Open source verification: Datadog has open-sourced the ptrace-based and eBPF agent, allowing clients and the security community to verify its safety and effectiveness themselves, fostering transparency and trust in our solution.

Reviewer: ```suggestion
- Open source verification: Datadog has open-sourced the ptrace-based and eBPF Agent, allowing clients and the security community to verify its safety and effectiveness themselves, fostering transparency and trust in the solution.
```

{{% /collapse-content %}} 

## eBPF-less agent setup

> You can set up the eBPF-less agent on various platforms, including Docker and Linux hosts.

Reviewer: ```suggestion
You can set up the eBPF-less Agent on various platforms, including Docker and Linux hosts.
```

> This section covers Docker and Linux hosts. For steps on setting up an Amazon fargate environment where eBPF is disabled, see [AWS Fargate Configuration Guide for Datadog Security][3].

Reviewer: ```suggestion
This section covers Docker and Linux hosts. For steps on setting up an Amazon Fargate environment where eBPF is disabled, see [AWS Fargate Configuration Guide for Datadog Security][3].
```

  --cgroupns host \
  --pid host \
  --security-opt apparmor:unconfined \
  --cap-add SYS_ADMIN \
  --cap-add SYS_RESOURCE \
  --cap-add SYS_PTRACE \
  --cap-add NET_ADMIN \
  --cap-add NET_BROADCAST \
  --cap-add NET_RAW \
  --cap-add IPC_LOCK \
  --cap-add CHOWN \
  -v /var/run/docker.sock:/var/run/docker.sock:ro \
  -v /proc/:/host/proc/:ro \
  -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
  -v /etc/passwd:/etc/passwd:ro \
  -v /etc/group:/etc/group:ro \
  -v /:/host/root:ro \
  -v /sys/kernel/debug:/sys/kernel/debug \
  -v /etc/os-release:/etc/os-release \
  -e DD_COMPLIANCE_CONFIG_ENABLED=true \
  -e DD_COMPLIANCE_CONFIG_HOST_BENCHMARKS_ENABLED=true \
  -e DD_RUNTIME_SECURITY_CONFIG_ENABLED=true \
  -e DD_RUNTIME_SECURITY_CONFIG_REMOTE_CONFIGURATION_ENABLED=true \
  -e DD_RUNTIME_SECURITY_CONFIG_EBPFLESS_ENABLED=true \
  -e HOST_ROOT=/host/root \
  -e DD_API_KEY=<API KEY> \
  gcr.io/datadoghq/agent:7
```
{{% /tab %}}

{{% tab "Linux host" %}}
To install the Agent to a Linux host, use the following install script to install the custom build:

```shell
DD_API_KEY=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX DD_SITE="datadoghq.com" \
DD_RUNTIME_SECURITY_CONFIG_ENABLED=true \

> DD_REPO_URL=datad0g.com \

Reviewer 2: This was needed to retrieve the custom build, now we can remove it

bash -c "$(curl -L https://install.datadoghq.com/scripts/install_script_agent7.sh)"
```

Next, modify the `/etc/datadog-agent/system-probe.yaml` file to enable CWS and eBPF-less mode as follow:

{{< code-block lang="java" filename="system-probe.yaml" disable_copy="false" collapsible="true" >}}
runtime_security_config:
  enabled: true
  ebpfless:
    enabled: true
{{< /code-block >}}

> Alternartively, to manually install the `.deb/.rmp` provided custom build packages, modify the `/etc/datadog-agent/system-probe.yaml` file to enable CWS and eBPF-less mode as follow:

Reviewer: ```suggestion
Alternatively, to manually install the `.deb/.rmp` provided custom build packages, modify the `/etc/datadog-agent/system-probe.yaml` file to enable CWS and eBPF-less mode as follows:
```

{{< code-block lang="java" filename="system-probe.yaml" disable_copy="false" collapsible="true" >}}
runtime_security_config:
  enabled: true
  ebpfless:
    enabled: true
{{< /code-block >}}
{{% /tab %}}

{{< /tabs >}}

> ## Deploy eBPF-less agent

Reviewer: ```suggestion
## Deploy eBPF-less Agent
```

> Ensure you perform the following configuration requirements before deploying the agent:

Reviewer: ```suggestion
Ensure you perform the following configuration requirements before deploying the Agent:
```

1. Customize the [Agent Installation Instructions][5] before proceeding with the installation.

> 2. Install/update the agent with CSM enabled. For steps, see [Setting up Cloud Security Management on the Agent][4]. 

Reviewer: ```suggestion
2. Install/update the Agent with CSM enabled. For steps, see [Setting up Cloud Security Management on the Agent][4]. 
```

3. Specify additional configurations from the previous **eBPF-less agent setup** sections to install the custom version and enable eBPF-less mode.

## Verify setup

> To validate your agent installation and setup, connect to your Linux host or Docker container and run:

Reviewer: ```suggestion
To validate your Agent installation and setup, connect to your Linux host or Docker container and run:
```

```shell
sudo /opt/datadog-agent/embedded/bin/system-probe config|grep -A 1 ebpfless
```

You should see the output:

```
  ebpfless:
    enabled: true
```

> ## Set up application tracing with eBPF-less agent

Reviewer: ```suggestion
## Set up application tracing with eBPF-less Agent
```

> After the eBPF-less agent is installed and set up to use the eBPF-Free mode, you can set up how your application is traced. This section provides you two different methods:

Reviewer: ```suggestion
After the eBPF-less Agent is installed and set up to use the eBPF-Free mode, you can set up how your application is traced. This section provides you two different methods:
```

- **Wrap mode:** (Recommended) In this mode, your application is launched by the Datadog wrapper that traces it from the beginning using ptrace.
  - All spawned children are traced also.
  - A seccomp profile is applied to drastically reduce the ptracing overhead.
- **Attach mode:** In this mode, you can specify a list of PIDs to attach to your application processes. This should be done quickly because your application is not ptraced until this is done. 
  - In this mode, a seccomp profile cannot be applied. Consequently, there is a small amount ptracing overhead.

> Both modes use the **cws-instrumentation** binary packaged with the datadog agent, and located at `/opt/datadog-agent/embedded/bin/cws-instrumentation`.

Reviewer: ```suggestion
Both modes use the **cws-instrumentation** binary packaged with the Datadog Agent, and located at `/opt/datadog-agent/embedded/bin/cws-instrumentation`.
```

<div class="alert alert-info">

> This tracer communicates with system-probe (part of the Datadog agent) on localhost using port 5678. The system-probe address can be configured with the <code>--probe-addr=host:port</code> cws-instrumentation option. The server-side address can be updated through the runtime_security_config.ebpfless.socket option of the <code>/etc/datadog-agent/system-probe.yaml</code> agent config file.

Reviewer: ```suggestion
This tracer communicates with system-probe (part of the Datadog Agent) on localhost using port 5678. The system-probe address can be configured with the <code>--probe-addr=host:port</code> cws-instrumentation option. The server-side address can be updated through the runtime_security_config.ebpfless.socket option of the <code>/etc/datadog-agent/system-probe.yaml</code> Agent config file.
```

</div>

{{< tabs >}}
{{% tab "Wrap mode" %}}
In wrap mode, the Datadog wrapper launchs the application. Here is an example:

```shell
sudo /opt/datadog-agent/embedded/bin/cws-instrumentation trace -- /usr/bin/your_application
```

If your application runs as non-root, specify the uid/gid as numeric values:

```shell
sudo /opt/datadog-agent/embedded/bin/cws-instrumentation trace --uid 100 --gid 100 -- /usr/bin/your_application
```

> <div class="alert alert-info">An application won’t start until cws-instrumentation has initialized its connection with the Datadog agent.</div>

Reviewer: ```suggestion
<div class="alert alert-info">An application won’t start until cws-instrumentation has initialized its connection with the Datadog Agent.</div>
```

# Start the service
start() {
        echo "Starting my app"
        /opt/datadog-agent/embedded/bin/cws-instrumentation trace -- /usr/bin/myapp &
}

# Stop the service
stop() {
       echo "Stopping my app"
	pkill -f /usr/bin/myapp
}

### main logic ###
case "$1" in
  start)
        start
        ;;
  stop)
        stop
        ;;
  restart)
        stop
        start
        ;;
  *)
        echo $"Usage: $0 {start|stop|restart}"
        exit 1
esac

exit 0
```

#### Docker

For Docker application deployments, you should modify your Dockerfile to wrap your application like this:

```shell
FROM gcr.io/datadoghq/agent:7 AS datadogagent

FROM ubuntu:latest

COPY --from=datadogagent /opt/datadog-agent/embedded/bin/cws-instrumentation .

ENTRYPOINT ["/cws-instrumentation", "trace", "--"]

CMD ["/bin/bash", "-c", "while true; do sleep 1; echo my app is running; done"]
```

When running your docker application, it’s important to give it an additional capability by adding `--cap-add=SYS_PTRACE` to your `docker run` command.

You also have to connect the container to Datadog on port 5678 by doing one of the following:

- Launch both containers with the `--network` host option.

> - Use the [Docker network][6] feature to run both containers on the same bridge network.

Reviewer: Noting that this link is broken in the staging preview.

Restart=on-failure

[Install]
WantedBy=multi-user.target
```

#### Linux sysvinit service

If you already have an init script, here is an example of how to integrate the tracer using a new sysvinit service:

```shell
#!/bin/sh
set -e
### BEGIN INIT INFO
# Provides:           dd_tracing_my_app
# Required-Start:     $network
# Required-Stop:      $network
# Default-Start:      2 3 4 5
# Default-Stop:       0 1 6
# Short-Description:  Datadog tracing of my application
# Description: Datadog tracing of my application
### END INIT INFO

# Start the service
start() {
        echo "Starting tracing my app"
        /opt/datadog-agent/embedded/bin/cws-instrumentation trace $(for pid in $(pidof myapp); do echo --pid $pid; done) &
}

# Stop the service
stop() {
       echo "Stopping my app"
	pkill -f /opt/datadog-agent/embedded/bin/cws-instrumentation
}

### main logic ###
case "$1" in
  start)
        start
        ;;
  stop)
        stop
        ;;
  restart)
        stop
        start
        ;;
  *)
        echo $"Usage: $0 {start|stop|restart}"
        exit 1
esac

exit 0
```

#### Docker

> To attach the wrapper to a Docker image running an application, use this following Dockerfile:

Reviewer: ```suggestion
To attach the wrapper to a Docker image running an application, use the following Dockerfile:
```

```shell

> FROM gcr.io/datadoghq/agent:latest

Reviewer 2: It should be the same, but to stay consistent with the other parts I would replace the `latest` tag by `7`

ENTRYPOINT ["/opt/datadog-agent/embedded/bin/cws-instrumentation", "trace", "--pid", "$PID"]
```

Next, provide the host PID for connecting to Docker as an environment variable.

To attach to an application, you’ll need the following:

- When running the Docker application, add the required capability by including `--cap-add=SYS_PTRACE` to your `docker run` command.
- Ensure the application container can reach the Datadog container on port 5678 using one of the following methods:
  - Launch both containers with the `--network` host option.

>   - Use the [Docker network][6] feature to run both containers on the same bridge network.

Reviewer: Same comment as above re: the missing destination link.

> - To ensure the application container is running on the host pid (just as the Datadog agent does), add these options: `--cgroupns host --pid host`.

Reviewer: ```suggestion
- To ensure the application container is running on the host pid (just as the Datadog Agent does), add these options: `--cgroupns host --pid host`.
```

        # [...]
        - name: system-probe
          image: "datadog/agent-dev:7-55-1-beta-host-ebpfless-v0"
          # [...]
          env:
            - name: DD_RUNTIME_SECURITY_CONFIG_EBPFLESS_ENABLED
              value: "true"
        # [...]
        - name: security-agent
          image: "datadog/agent-dev:7-55-1-beta-host-ebpfless-v0"
      # [...]
      initContainers:
        - name: init-volume
          image: "datadog/agent-dev:7-55-1-beta-host-ebpfless-v0"
        # [...]
        - name: init-config
          image: "datadog/agent-dev:7-55-1-beta-host-ebpfless-v0"
        # [...]
        - name: seccomp-setup
          image: "datadog/agent-dev:7-55-1-beta-host-ebpfless-v0"
{{< /code-block >}}
{{% /tab %}}
{{% tab "Linux host" %}}
To install the agent to a Linux host, use the following install script to install the custom build:

```shell
DD_API_KEY=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX DD_SITE="datadoghq.com" \
DD_RUNTIME_SECURITY_CONFIG_ENABLED=true \
DD_REPO_URL=datad0g.com \
DD_AGENT_MINOR_VERSION=55.1~beta~host~ebpfless~v0-1 \
DD_AGENT_DIST_CHANNEL=beta bash -c "$(curl -L https://install.datadoghq.com/scripts/install_script_agent7.sh)"
```

> Alternartively, to manually install the `.deb/.rmp` provided custom build packages, modify the `/etc/datadog-agent/system-probe.yaml` file to enable CWS and eBPF-free mode as follow:

Author: okay. I added it to the install script section.

   [Service]
   ExecStart=/opt/datadog-agent/embedded/bin/cws-instrumentation trace -- /usr/bin/myapp
   Restart=on-failure

   [Install]
   WantedBy=multi-user.target
```

#### Linux sysvinit service

If you already have an init script, here is a simple example of the required changes:

```shell
#!/bin/sh
set -e
### BEGIN INIT INFO
# Provides:           my_app
# Required-Start:     $network
# Required-Stop:      $network
# Default-Start:      2 3 4 5
# Default-Stop:       0 1 6
# Short-Description:  My application
# Description: My application
### END INIT INFO

# Start the service
start() {
        echo "Starting my app"
        /opt/datadog-agent/embedded/bin/cws-instrumentation trace -- /usr/bin/myapp &
}

# Stop the service
stop() {
       echo "Stopping my app"
	pkill -f /usr/bin/myapp
}

### main logic ###
case "$1" in
  start)
        start
        ;;
  stop)
        stop
        ;;
  restart)
        stop
        start
        ;;
  *)
        echo $"Usage: $0 {start|stop|restart}"
        exit 1
esac

exit 0
```

#### Docker

For Docker application deployments, you should modify your Dockerfile to wrap your application like this:

```shell

> FROM datadog/agent-dev:7-55-1-beta-host-ebpfless-v0 AS datadogagent

Reviewer 2: replace version with datadog/agent:latest

       command:
         - "/opt/datadog-agent/embedded/bin/cws-instrumentation"
         - "setup"
         - "--cws-volume-mount"
         - "/cws-instrumentation-volume"
       volumeMounts:
         - name: cws-instrumentation-volume
           mountPath: "/cws-instrumentation-volume"
       securityContext:
         runAsUser: 0
     containers:
     - name: "my_app"
       image: "my_image:latest"
       command:
         - "/cws-instrumentation-volume/cws-instrumentation"
         - "trace"
         - "--"
         - "my_entrypoint"
       volumeMounts:
         - name: cws-instrumentation-volume
           mountPath: "/cws-instrumentation-volume"
           readOnly: true
     - name: datadog-agent
       image: public.ecr.aws/datadog/agent-dev:7-55-1-beta-host-ebpfless-v0
       env:
         - name: DD_API_KEY
           value: "<DD_API_KEY>"
         - name: DD_RUNTIME_SECURITY_CONFIG_ENABLED
           value: "true"
         - name: DD_RUNTIME_SECURITY_CONFIG_EBPFLESS_ENABLED
           value: "true"
     volumes:
       - name: cws-instrumentation-volume
     serviceAccountName: datadog-agent
     shareProcessNamespace: true

> ```

Reviewer 2: kube section to be removed

{{% /tab %}}

{{% tab "Attach mode" %}}
The wrap mode is recommended because the attach mode has the following limitations: 

- It misses all initialization made by the application until Datadog attaches to it.

> - More performance overhead.

Reviewer 2: Maybe we can add the why again here: because when attaching we cannot setup any seccomp profile

[Install]
WantedBy=multi-user.target
```

#### Linux sysvinit service

If you already have an init script, here is an example of how to integrate the tracer using a new sysvinit service:

```shell
#!/bin/sh
set -e
### BEGIN INIT INFO
# Provides:           dd_tracing_my_app
# Required-Start:     $network
# Required-Stop:      $network
# Default-Start:      2 3 4 5
# Default-Stop:       0 1 6
# Short-Description:  Datadog tracing of my application
# Description: Datadog tracing of my application
### END INIT INFO

# Start the service
start() {
        echo "Starting tracing my app"
        /opt/datadog-agent/embedded/bin/cws-instrumentation trace $(for pid in $(pidof myapp); do echo --pid $pid; done) &
}

# Stop the service
stop() {
       echo "Stopping my app"
	pkill -f /opt/datadog-agent/embedded/bin/cws-instrumentation
}

### main logic ###
case "$1" in
  start)
        start
        ;;
  stop)
        stop
        ;;
  restart)
        stop
        start
        ;;
  *)
        echo $"Usage: $0 {start|stop|restart}"
        exit 1
esac

exit 0
```

#### Docker

To attach the wrapper to a Docker image running an application, use this following Dockerfile:

```shell

> FROM datadog/agent-dev:7-55-1-beta-host-ebpfless-v0

Reviewer 2: same here, datadog/agent:latest :)
